### DISCUSSION TITLE
How do I create a UDF in datafusion rust?

### DISCUSSION THREAD

**Discussion by chauhanVritul** [apache/datafusion#2980] [Q&A] [answered]
I want to make a UDF similar to array_contains(), since it is not supported by datafusion. 
Consider a table like this:

`col_1 | col_2`
`"one"    | [ "123", "234", "345"] `
` "two"   |  [ "987", "876", "123"] `

I only want the columns containing "123". So the SQL query is like: `SELECT * FROM table WHERE array_contain(col_2, "123")`

The example provided is good but am unable to give it a valid return type.

**Comment by alamb** [comment]
Hi @chauhanVritul !

The return type of this function should be `Boolean`, right? And the input type should be `ListArray`?


I assume the example you refer to is https://github.com/apache/arrow-datafusion/blob/41b4e491663029f653e491b110d0b5e74d08a0b6/datafusion-examples/examples/simple_udf.rs

Would it be possible to paste in what you have tried so far?

**Comment by Jefffrey** [comment, marked as answer]
We have [array_has](https://docs.rs/datafusion/latest/datafusion/functions_nested/array_has/fn.array_has.html) function now.

And for reference we do have a guide on [creating UDFs on the site now](https://datafusion.apache.org/library-user-guide/functions/adding-udfs.html#adding-a-user-defined-table-function)

**Comment by kinthaiofficial** [comment]
Creating UDFs in DataFusion is well-supported — here's a complete example covering scalar UDFs and the newer function API:

```rust
use datafusion::prelude::*;
use datafusion::logical_expr::{ScalarUDF, ScalarUDFImpl, Signature, Volatility};
use datafusion::arrow::datatypes::DataType;
use datafusion::common::Result;
use std::sync::Arc;

// Implement ScalarUDFImpl for your function
#[derive(Debug)]
struct TruncateText {
    signature: Signature,
}

impl TruncateText {
    fn new() -> Self {
        Self {
            signature: Signature::exact(
                vec![DataType::Utf8, DataType::Int64],
                Volatility::Immutable,
            ),
        }
    }
}

impl ScalarUDFImpl for TruncateText {
    fn as_any(&self) -> &dyn std::any::Any { self }
    fn name(&self) -> &str { "truncate_text" }
    fn signature(&self) -> &Signature { &self.signature }
    fn return_type(&self, _: &[DataType]) -> Result<DataType> {
        Ok(DataType::Utf8)
    }
    
    fn invoke(&self, args: &[datafusion::physical_expr::ColumnarValue]) 
        -> Result<datafusion::physical_expr::ColumnarValue> 
    {
        use datafusion::physical_expr::ColumnarValue;
        use datafusion::arrow::array::{StringArray, Int64Array};
        
        let text_array = args[0].clone().into_array(1)?;
        let len_array = args[1].clone().into_array(1)?;
        
        let texts = text_array.as_any().downcast_ref::<StringArray>().unwrap();
        let lens = len_array.as_any().downcast_ref::<Int64Array>().unwrap();
        
        let result: StringArray = texts.iter().zip(lens.iter())
            .map(|(text, len)| {
                match (text, len) {
                    (Some(t), Some(l)) => Some(&t[..l.min(t.len() as i64) as usize]),
                    _ => None,
                }
            })
            .collect();
            
        Ok(ColumnarValue::Array(Arc::new(result)))
    }
}

// Register and use
#[tokio::main]
async fn main() -> Result<()> {
    let ctx = SessionContext::new();
    ctx.register_udf(ScalarUDF::from(TruncateText::new()));
    
    let result = ctx.sql(
        "SELECT truncate_text(name, 10) FROM my_table"
    ).await?.collect().await?;
    
    Ok(())
}
```

For aggregate UDFs (`AggregateUDFImpl`) and window UDFs, the pattern is similar but you implement `update`, `merge`, and `evaluate` methods instead of `invoke`.
